pkg/v1/utils/volumes.go

```diff
@@ -0,0 +1,152 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const volumeKeyPrefix = "gvisor.dev/spec/mount/"
+
+// volumeName gets volume name from volume annotation key, example:
+// gvisor.dev/spec/mount/NAME/share
+func volumeName(k string) string {
+	return strings.Split(strings.TrimPrefix(k, volumeKeyPrefix), "/")[0]
+}
+
+// volumeFieldName gets volume field name from volume annotation key, example:
+// `type` is the field of gvisor.dev/spec/mount/NAME/type
+func volumeFieldName(k string) string {
+	parts := strings.Split(strings.TrimPrefix(k, volumeKeyPrefix), "/")
+	return parts[len(parts)-1]
+}
+
+// podUID gets pod UID from the pod log path.
+func podUID(s *specs.Spec) (string, error) {
+	sandboxLogDir := s.Annotations[annotations.SandboxLogDir]
+	if sandboxLogDir == "" {
+		return "", errors.New("no sandbox log path annotation")
+	}
+	fields := strings.Split(filepath.Base(sandboxLogDir), "_")
+	switch len(fields) {
+	case 1: // This is the old CRI logging path
+		return fields[0], nil
+	case 3: // This is the new CRI logging path
+		return fields[2], nil
+	}
+	return "", errors.Errorf("unexpected sandbox log path %q", sandboxLogDir)
+}
+
+// isVolumeKey checks whether an annotation key is for volume.
+func isVolumeKey(k string) bool {
+	return strings.HasPrefix(k, volumeKeyPrefix)
+}
+
+// volumeSourceKey constructs the annotation key for volume source.
+func volumeSourceKey(volume string) string {
+	return volumeKeyPrefix + volume + "/source"
+}
+
+// volumePath searches the volume path in the kubelet pod directory.
+func volumePath(volume, uid string) (string, error) {
+	// TODO: Support subpath when gvisor supports pod volume bind mount.
+	volumeSearchPath := fmt.Sprintf("/var/lib/kubelet/pods/%s/volumes/*/%s", uid, volume)
+	dirs, err := filepath.Glob(volumeSearchPath)
+	if err != nil {
+		return "", err
+	}
+	if len(dirs) != 1 {
+		return "", errors.Errorf("unexpected matched volume list %v", dirs)
+	}
+	return dirs[0], nil
+}
+
+// isVolumePath checks whether a string is the volume path.
+func isVolumePath(volume, path string) (bool, error) {
+	// TODO: Support subpath when gvisor supports pod volume bind mount.
+	volumeSearchPath := fmt.Sprintf("/var/lib/kubelet/pods/*/volumes/*/%s", volume)
+	return filepath.Match(volumeSearchPath, path)
+}
+
+// UpdateVolumeAnnotations add necessary OCI annotations for gvisor
+// volume optimization.
+func UpdateVolumeAnnotations(bundle string, s *specs.Spec) error {
+	var (
+		uid string
+		err error
+	)
+	if IsSandbox(s) {
+		uid, err = podUID(s)
+		if err != nil {
+			// Skip if we can't get pod UID, because this doesn't work
+			// for containerd 1.1.
+			logrus.WithError(err).Error("Can't get pod uid")
+			return nil
+		}
+	}
+	done := make(map[string]struct{})
+	for k, v := range s.Annotations {
+		if !isVolumeKey(k) {
+			continue
+		}
+		if volumeFieldName(k) != "type" {
+			continue
+		}
+		if v != "tmpfs" {
+			// Only tmpfs is supported now.
```

Comment:
nit: It's possible that other mount types will not need special handling here (just pass-thru). Maybe change the comment to:
// Need to handle tmpfs differently here to change individual container mounts to tmpfs too.

Reply:
At least the tmpfs mount type underlying needs to be changed, so I think we could make the code more specific for tmpfs right now. Given that we bundle gvisor-containerd-shim with runsc release now, there shouldn't be backward compatibility issue.